aip/general/0121.md

```diff
@@ -81,6 +81,26 @@ In an API with a stateless protocol, the server has the responsibility for
 persisting data, which may be shared between multiple clients, while clients
 have sole responsibility and authority for maintaining the application state.
 
+### Cyclic References
```

Comment:
Can a cyclic graph be designed in such a way that would not impact the complexity of the design. For example with back references that were tagged as "do-not-follow"?

Reply:
I don't think do-not-follow would address the issue: the primary problem is with it's impact on ordering of creation and deletion, vs something like resource traversal (which do-not-follow would help with).

For traversal I don't think you'd need an explicit marker anyway: you could store the set of visited nodes and reference those so you don't double-visit.

Reply:
What are we giving up not allowing cyclic dependencies? 
Do we know of use cases where such restrictions would prevent an implementation?

Reply:
> Do we know of use cases where such restrictions would prevent an implementation?

Could you clarify what implementation you're referring to? 

This does not prevent an API from being implemented, but it does make it very difficult for a declarative client to generate itself. One example is Terraform, for the resource ordering problem I've outlined in the description (a declarative client has to do a three-phase imperative apply to get the desired state, and plan resolvers often do not have the facilities to do so).